### ISSUE TITLE
Code generation that deduplicates validation logic

### ISSUE THREAD

**Issue by ggoodman** [ExodusMovement/schemasafe#140]
## Use-case

Imagine you're building validator codecs for a set of inter-related schemas that define `$ref`s on each other. The (awesome) code generation results in these being treated as totally isolated IIFE-enclosed validators. These validators duplicate the validation logic of all referenced schemas.

## Request

It would be really interesting to have a facility whereby a set of inter-related schemas could be compiled as a unit and thereby produced smaller validation logic by referencing each other appropriately. I suspect that this would require major changes to how errors are collected as we would now have `N` validator functions instead of `1` on which errors may be added.

Anyway, schemasafe is a _fantastic_ bit of work. Really appreciate the rigour, speed and comprehensiveness. I'm using it with https://github.com/ggoodman/json-schema-to-dts to build codegen typed validation codecs.

**Comment by ChALkeR** [comment]
Sorry for the late reply on this.

This is an interesting question.

One of the core reasons for using separate functions is that they might call each other:
```json
{
  "$defs": {
    "a": {
      "type": "object",
      "properties": {
        "value": { "$ref": "#/$defs/b" }
      }
    },
    "b": {
      "type": "array",
      "items": { "$ref": "#/$defs/a" }
    }
  },
  "$ref": "#/$defs/a"
}
```

Even for schemas without cyclic refs, if a ref is used multiple times, that can lead to more compact code.

That said, in some situations inlining refs could be both possible and beneficial (i.e. lead to smaller generated code size).

**Comment by ChALkeR** [comment]
Ah, disregard the previous comment, I misunderstood.

Compiling several schemas into one object to avoid duplication is an interesting request, this should definitely be viable.
Will require some some API changes though, I will need to think about this.

**Comment by ChALkeR** [comment]
https://github.com/ExodusMovement/schemasafe/compare/chalker/multi-schema implements this as a draft
The API will likely be different though
Also, untested yet

**Comment by ChALkeR** [comment]
This also works now:
```js
const schemas = [
  {
    $id: "https://example.com/a",
    "type": "object",
    "properties": {
      "value": { "$ref": "https://example.com/b" }
    }
  },
  {
    $id: "https://example.com/b",
    "type": "array",
    "items": { "$ref": "https://example.com/a" }
  },
  {}
]
const validates = validators(schemas)
console.log(validates[1]([]))
console.log(validates.toModule())
```

The API will likely change though.

**Comment by ChALkeR** [comment]
The API has changed to `validate(schemas, { multi: true })` instead of `validators(schemas)`.
That'll likely be the one that will get landed.

**Comment by ChALkeR** [comment]
Implemented in https://github.com/ExodusMovement/schemasafe/commit/12727746f92a2f649a72d23d2c83a44864b723d4, now both `parser()` and `validator()` have a new `{ multi: true }` option that allows compiling an array of schemas at once to a single module.

See example in tests.

Not properly documented for now.

**Comment by ChALkeR** [comment]
Released in [v1.0.0-rc.9](https://github.com/ExodusMovement/schemasafe/releases/tag/v1.0.0-rc.9)
